### PR TITLE
lnd: ensure that LND won't start in native SQL mode if it has any KV invoices 

### PR DESCRIPTION
**Pull request by bhandras**
This PR is a temporary measure for 0.18 as we don't plan to add invoice migration yet for native SQL invoice DB.

### CODE REVIEW

File: config_builder.go [hunk 1]
```diff
@@ -1022,6 +1022,34 @@ func (d *DefaultDatabaseBuilder) BuildDatabase(
 
 	// Instantiate a native SQL invoice store if the flag is set.
 	if d.cfg.DB.UseNativeSQL {
+		// KV invoice db resides in the same database as the graph and
+		// channel state DB. Let's query the database to see if we have
+		// any invoices there. If we do, we won't allow the user to
+		// start lnd with native SQL enabled, as we don't currently
+		// migrate the invoices to the new database schema.
+		invoiceSlice, err := dbs.GraphDB.QueryInvoices(
+			context.Background(), invoices.InvoiceQuery{
```
Comment (ellemouton) [inline config_builder.go:1031]: i think you can instead use the context that is already available

File: lncfg/db.go [hunk 2]
```diff
@@ -122,13 +122,24 @@ func DefaultDB() *DB {
 // Validate validates the DB config.
 func (db *DB) Validate() error {
 	switch db.Backend {
-	case BoltBackend, SqliteBackend:
+	case BoltBackend:
+		if db.UseNativeSQL {
+			return fmt.Errorf("cannot use native SQL with bolt " +
+				"backend")
```
Comment (ellemouton) [inline lncfg/db.go:128]: 🙏

**Review by ellemouton — APPROVED** [review]
Looking good!

One comment about re-using the passed in context instead.

Then as mentioned offline: realised that we dont do this check the other way around (ie it is possible to add native sql invoice & then remove the flag and it will go back to adding kvdb invoices). But Andras pointed out that this is already the case with any backend change we have today. So potentially not needed to be that defensive especially cause in future when we add migration scripts, then we do want to allow this 👍

File: config_builder.go (hunk 1 above)
Comment (bhandras) [reply config_builder.go]: Fixes, thanks :)

**Review by bitromortac — APPROVED** [review]
LGTM :tada:. Only to clarify, it doesn't matter if the database contained any invoices that were deleted, we just require that it is empty?

**Comment by bhandras** [comment]
> LGTM 🎉. Only to clarify, it doesn't matter if the database contained any invoices that were deleted, we just require that it is empty?

Good question! As deleted invoices are completely removed from the KV invoice DB, it'd be fine. We could be more strict by not even caring about the invoice DB, but only about whether this is a first ever start of the node. I believe the current approach is quite liberal in that it allows "old" nodes to also use the native schema (given no invoices).

**Comment by bhandras** [comment]
> LGTM! tACK 🚀
> 
> One comment about re-using the passed in context instead.
> 
> Then as mentioned offline: realised that we dont do this check the other way around (ie it is possible to add native sql invoice & then remove the flag and it will go back to adding kvdb invoices). But Andras pointed out that this is already the case with any backend change we have today. So potentially not needed to be that defensive especially cause in future when we add migration scripts, then we do want to allow this 👍

Thanks for the testing and discussion! :) I agree that we could be more strict about preconditions, see also my other comment below. I'm also open to changing things up if you both think more strict checking is better.

**Pull request merged**